### ISSUE TITLE
General libcxx8 issues

### ISSUE THREAD

**Issue by chuggafan** [LADSoft/OrangeC#355]
This branch https://github.com/chuggafan/OrangeC/commit/54b1353e3adb2098865401b636c4ad020c8e714d is just a pure libcxx8 replacement of our current libcxx work, with threading enabled, and I fixed the has_feature thing to work better with how libcxx does it and the way the c preprocessor works. 

OCC has an ICE at <utility> line 901, no idea why, and a bunch of issues you'll see that I've made will pop up (all 4 of the recent ones), I removed the "custom" news and am using the libcxx one, it's the same thing, just inbuilt to the library.

other than that, not much to say other than we've only scratched the very tip of this iceberg of issues so far.

**Comment by LADSoft** [comment]
I checked in changes for the recent issues; other than to validate that my smaller example of the problems in #350 works I didn't try to actually use the results of #350 yet.   I figured I will get the chance when I start looking at this lol!

**Comment by chuggafan** [comment]
I've pushed an update that should fix the majority of the "simpler" problems thanks to your work, it actually goes through most of the stuff.

Main problem seems that libcxx likes to treat <charconv> as c++14 even though it's c++17 so that bugs out... Still got some weird other issues going on as well.

Also, it seems like the compiler accepts ```new``` with multiple parameters, however any ```delete``` with multiple parameters is an error, this should be fixed, I do believe that it's simple enough as well, though, I think just removing the extra check (just making sure first arg is a pointer) would do the trick (though that would allow for BS operator deletes, and BS operator news are already allowed though...). That's one bug...

**Comment by LADSoft** [comment]
thanks.  sorry for the delay, been copying my files to a new computer...

A simple compile of <version> results in 96 errors now...   including the ones you indicated above.   Many of them seem related to evaluating using statements too early...   I'll have to look into that.

The compiler is still C++14 BTW, haven't done the C++17 work yet.

I really need to make time for that constexpr issue at some point, the code generation for some supposedly constexpr values is surreal!

BTW I fixed a crash here before I got into the new computer last night: 61226c15c07a48dab6ae3f4d34670bdeae0dd21d.   Apparently some code I put in a long time ago never got triggered until now lol!

**Comment by LADSoft** [comment]
i grabbed at the low-hanging fruit and added __is_abstract to the compiler builtins.   Took out 18 errors - down to 78 now.   I'll look at it closer tomorrow...

**Comment by LADSoft** [comment]
typeinfo.h does not seem to be defining bad_cast?

**Comment by LADSoft** [comment]
I'm down to 34 issues on the vector compile.   Will get back to it later...

**Comment by LADSoft** [comment]
i got vector to compile but now the compiler basically doesn't compile more complex files...   rather than fix it immediately I've started on map...   But I'm having a problem with the declaration of **__make_tree_node_types** in **__tree**.   Any advice you could give would be welcome...   

basically I'm having a hard time understanding why the expansion of its 'type' typedef should result in the __tree_node_types from line 662.   I know it should but I don't know the rationale...   So I can't keep OCC from selecting the definition at line 659

**Comment by chuggafan** [comment]
I wasn't able to see any of the __tree stuff, but I did see that there's a bunch of issues in cmath due to fpclassify etc. being defined as macros instead of actual functions, which should be fixable.... <memory> seems to have a bunch of issues as well, and <type_traits> is the big one for issues.... *sigh* I'll get to it soon-ish, I've been focusing on other things over the weekend...

**Comment by LADSoft** [comment]
I got past that...

I've been fixing bugs.   There are rather a lot.   But I'm down to three left in map header... 

I can't release the compiler on the main branch because it needs fixing now that I've been working on it,  about the only thing it is good for is bug cleanup for libcxx right now lol!   but if you want to see where things are for you with the current work I can make another branch for the WIP. :)

**Comment by chuggafan** [comment]
I'm always happy to checkout the latest work... TBH I should also mention that LIBCXX has a testing method for it that we should integrate that tests the whole thing, but that'll be a pain that we might need to save for after we get it to simply compile...

**Comment by LADSoft** [comment]
that sounds good.   After we get the libcxx working and can compile real programs again I definitely want to do that.

I added a branch libcxx8-compiler with the current work...   I cleaned up a lot of issues in type-traits and utility while working on vector...

**Comment by chuggafan** [comment]
So I fixed one of the problems I found (the std::bad_cast stuff, which got some stuff in any.cpp found instead, yay that's improvement) and one of the roadblocks we're on again is #248 where exception_pointer is unimplemented and until then I'm wary of adding a libcxx compatible OrangeC header file.... after that we still need to make our <atomic> implementation compatible enough for std::atomic support, but I guess that can come a bit later (as we currently do not have it anyways...)

**Comment by LADSoft** [comment]
I had compeletely forgetten about #248.   I'm going to finish the work with std::map since it is close then I will work on that :)

**Comment by chuggafan** [comment]
Oh, as an aside, the problems found in \<cmath\> is actually caused by those functions that "aren't defined" being macros instead of pure functions, I also discovered that I was building libcxx wrong, you're supposed to have defined \_LIBCPP\_BUILDING\_LIBRARY if you're compiling the C++ files, also, it seems that \<string\> is now causing an access violation... I'll investigate it more at another time, but it seems to be there... I'll load up my latest work (goddamnit my git skills are dumb) in the meantime though... I did fix type_info thankfully. so yay on that...

**Comment by LADSoft** [comment]
i have to do a search on the internet any time I do anything in Git other than a simple checkin or branch switching lol!   Ok will look at std::string.   Thanks for looking a the type_info problem.

**Comment by LADSoft** [comment]
i checked in a new version.   compiles the vector, map, and string headers from the new library now.   I also fixed occ so that it would again compile the versions of those files in the old library.

Next I will get rid of the diagnostics and get code using those headers running.  Then will go on to #248.

**Comment by LADSoft** [comment]
oops need to turn off the CI builds for this branch I guess.

It needs more work too, will get back to you...

**Comment by chuggafan** [comment]
It's been a bit, I know you're busy, has this progressed any further recently? Anything you need help with that I should check out?

**Comment by LADSoft** [comment]
i kinda got sidetracked.   As I left it I think it was compiling stuff and I could get vectors to work, but I don't remember if maps were compiling or not.   At some point I had them compiling but they didn't work...

anyway...

i've most recently been sidetracked with #GitMenche suggesting to get things ready for another release.   And I've also been kinda sidetracked with a lot of personal stuff that came up...  anyway I have to finish up one more thing for a release then I can get back to this...

**Comment by LADSoft** [comment]
I will probably work on this again starting tomorrow.   By then the #348 should clear.   I'm going to defer the remaining issues that were found over the last few days until this weekend.

**Comment by GitMensch** [comment]
Just to recheck - is there any progress with this issue?

**Comment by chuggafan** [comment]
#366 is a legitimate issue that will get us a step further to actually acomplishing this if it gets fixed, however there's still a lot of problems either with exceptions or getting libcxx to cooperate with what we do and do not have before we can declare this war over.

**Comment by chuggafan** [comment]
Going to restart my quest for this issue to get closed eventually soon, I have no idea how far the libcxx-8 branch of the compiler has drifted from the master at this point, and so I'm going to be tackling issues from scratch especially since 366 was just completed.

The main problem points will of course be: exceptions (we need to re-work our entire std::exception stuff to deal with C++17 exception spec even if we're just running C++14 since libcxx expects it) and making the libcxx atomics work is going to be a ton of work on changing code so that there's a custom implementation. Threading should be able to be handled easily now, and the only problems that we should find other than exception are either internal compiler problems or libcxx expecting different things out of the C library than are given.

**Comment by chuggafan** [comment]
I've uploaded my current work at https://github.com/chuggafan/OrangeC/tree/libcxx8mods the main problems so far (algorithms.cpp specific):

in iterator for __wrap_iter base is somehow not defined despite being defined

In __tuple there's some definite magic going on because a template isn't working correctly when it should (I think because it's size_t and equal to 0 or some other stuff that it might have to do with constexpr evaluation, need to look into that)

In memory it's looking for redifinition for default value for types that are anonymous according to the compiler. I have no idea what thing the compiler is referring to and I can only imagine what arcane thing causes this.

There's a lot of different issues that are popping up, and those seem like good starts to me.

**Comment by LADSoft** [comment]
I'm going to work on a few other issues before I get back to this.   #366 wore me out!  Those should be clear in a week or two.    But I definitely want to prioritize it pretty high.  But I think I mentioned in #366 that I want to write a comprehensive dump of what is going on with the template handling code to a file somewhere to make diagnosing the issues a little  quicker...

and i agree restarting is a good idea :)

**Comment by LADSoft** [comment]
@chuggafan I was wondering if there was any reason you know of we couldn't go directly to libcxx 10 skipping 8 and 9?

**Comment by chuggafan** [comment]
There is absolutely none, only issue on my end is that it's harder to get JUST libcxx cloned from the git repos from LLVM.

**Comment by LADSoft** [comment]
going to go with 10....   rather than cloning the repo I'm downloading it as a src.tar.xz file.

**Comment by LADSoft** [comment]
i'm back to working on this...

**Comment by LADSoft** [comment]
the following currently doesn't work with libcxx10.   OCC on the libcxx branch doesn't currently compile it, but last time it did get compiled (with the commented pars uncommented) it produced incorrect output:

[s6.cpp.txt](https://github.com/LADSoft/OrangeC/files/5059521/s6.cpp.txt)

here is the libcxx 10 headers I am currently using, unzip them to "c:\libcxx 10"  then use the CLX10.BAT file in the orangec\src directory to install them.   If you want to switch back to libcxx 7 use CLX.BAT.

[libcxx10.zip](https://github.com/LADSoft/OrangeC/files/5059552/libcxx10.zip)

**Comment by chuggafan** [comment]
I'm now taking a look, so the main blocking problem is that the `using a = whatever;` construct is either not compiling, or is producing an incorrect value, am I understanding this correctly?

**Comment by chuggafan** [comment]
Ok, now that I have visual studio open and am actually analyzing things, CLX10.bat misses one key area: the source files to the includes, which are incredibly important to libcxx actually working for many parts of it. I'll get to working on the other stuff regarding actual-source-code-only issues for now, but I'm just going to throw that out there.

**Comment by LADSoft** [comment]
yeah the using a= did compile properly at one point but is crashing now; also when it compiled the results were sometimes right and sometimes wrong.   The actual results seemed to depend on previous results expanding the same using statement.

The reason the source code isn't included at this point is we can't compile it until we get the headers to compiler properly...   that is another task entirely...   I've been concentrating on things like vector and map which are mostly header-only for now.    The example I gave happened while compiling the 'map' header and trying to use the results...

**Comment by chuggafan** [comment]
So discovered a problem that is causing many of the issues when I'm personally trying compiles atm on my include_next branch (somehow I think it works in practice but not in testing so I'm not getting errors currently after a few modifications to the include files for standard conformance and the build).

On line 9586 of template.cpp the isstructured macro expands into something that is dereferencing illegal memory when attempting to obtain the rootType of a specialization. The specific file it seems to die on is __tuple. 

It seems to be shoving an invalid memory address into rootType, in this case, 0x00000085, however, the beast that is occparse is neither an easy location to find where something that got shoved into the nestedSearch stack and when it happened. __enable_if_tuple_size_imp is the specific symbol it dies on.

**Comment by LADSoft** [comment]
hm will see if I can figure out what is up with that...

**Comment by LADSoft** [comment]
I'm about ready to get back to this.   I was thinking, I was having to hack a lot of stuff together to try to get using statements to work reliably...   I'm thinking to pull a lot of that work back out and rewrite how using statements are done from scratch to make this better :smile:

**Comment by LADSoft** [comment]
I've rewritten how the templated using statements work and I think the new design is much cleaner.    Hope to have it tested out within a few days.

**Comment by LADSoft** [comment]
so this is coming along well, I got the new type alias statements working again and the rewrite got us past the issues I was having in June.    It did set me back a bit;  LIBCXX10 vectors and strings broke in a big way after the rewrite but I did finally get them to compile again.   

I have another issue with a complex template argument when compiling maps now but hope that I can get them working within the next few days.   I'm also working on the side to get it to compile libcxx8 again...   the test applications in the test directory currently work but when it comes to compiling the suite with itself some stuff is broken.

What I'm hoping for is that I can get the work back on the master branch so I can take a break and clean up #501...    well we'll see I guess...

**Comment by LADSoft** [comment]
making progress...  actually got it to find the correct constructor for a tuple which was a major undertaking in and of itself...   then I worked through some bugs with unmangling and error reporting and some other stuff.    Have to finish getting it to compile tuple_impl next then hopefully LIBCXX10 maps will fully compile.

**Comment by LADSoft** [comment]
last week i had to make a change which broke some stuff again...   this week I spent all week finding out that I have to support things like:

template <class ..._Tp>
using aa = some template<some type alias<decltype(some complicated template function(_Tp{})::type>::type...>

lol it is always an adventure!   I'll probably get past this one this weekend then we'll see where we stand...   Maybe if I'm lucky this will get me past maps!

**Comment by LADSoft** [comment]
ok so i got through tuples this weekend, back to working on maps.    Starting to see light at the end of this tunnel :smile:

**Comment by LADSoft** [comment]
so now my map demo compiles properly in libcxx10 finally (can't run it yet due to some changes to the mangled names); I'm working on fixing all the introduced bugs that make it so I cannot compile libcxx8.

**Comment by LADSoft** [comment]
it is coming along pretty well.    Got through the ostream header easily enough; working on building the C++ library files for libcxx 10.   About halfway through.   So far nothing monumental...

**Comment by LADSoft** [comment]
getting close on this.   Managed to compile the libcxx-10 c++ library files now.   created a bug while compiling the libcxx8 library though lol!   Along the way I had fixed static_assert() to work better (it wasn't asserting inside structure declarations) but it highlighted a problem.   

I'm going to vet the project against the libcxx-8 library one final time then hope to be able to start compiling the project with the libcxx 10 library this weekend.

**Comment by LADSoft** [comment]
still making progress, now I've compiled the RTL using LIBCXX 10 with the latest compiler.    It compiles, but there are some missing externals...   One thing I was surpised about is that you could use the 'default' keyword on a constructor outside of a class definition, and that would force the constructor to be instantiated...

**Comment by LADSoft** [comment]
so now the RTL compiles and I'm able to compile other C++ programs with it both with and without lscrtl.dll.   next up, I start compiling the test suite and when that works I'll compile the compiler and tools...

**Comment by chuggafan** [comment]
Currently being prevented from building locale.cpp when doing fullbuilds by two things: One: we don't have support for locale_win32, which if we just basically copied what MSVC does with locales we can piggyback and claim that we support that or we can do something else. Two:

```
Error(199)    C:\OrangeC\include\tuple(232):  Reference initialization of type 'std::__1::{unique}::__fake_bind' cannot be bound to type 'std::__1::{unique}::__fake_bind'
note:         C:\OrangeC\include\tuple(232):  Referenced in instantiation of 'std::__1::__tuple_leaf<unsigned int=0, __fake_bind&&, bool=0>::__tuple_leaf<__fake_bind, bool, >(__fake_bind&&)'
note:         C:\OrangeC\include\tuple(387):  Referenced in instantiation of 'std::__1::__tuple_impl<__tuple_indices<...unsigned int=0>, __fake_bind&&>::__tuple_impl<..., unsigned int=0, __fake_bind&&, ..., ..., _Tl, _Tl, >(__tuple_indices<...unsigned int=0>, __tuple_types<_Tl&&>, __tuple_indices<...>, __tuple_types<..._Tp>, _Tp&&)'
note:         C:\OrangeC\include\tuple(793):  Referenced in instantiation of 'std::__1::tuple<__fake_bind&&>::tuple<__fake_bind, bool=0, bool=0, >(__fake_bind&&)'
note:         C:\OrangeC\include\mutex(669):  Referenced in instantiation of 'std::__1::call_once<__fake_bind, ..._Args>(once_flag&, __fake_bind&&, void)'
note:         locale.cpp(658):  In template instantiation started here
```

Seems to me: instantiation of the type is not being treated equal to itself due to something about it being unique?

When investigating, it seems to be a problem I *THINK* with std::forward, I can't really tell.

**Comment by LADSoft** [comment]
yeah locale.cpp is a sticky wicket...   it is usually one of the first files to exhibit problems when I break something lol!   I've had problems with __fake_bind before but I'm not seeing it now.    as far as locale_win32 I didn't want to support the msvc way of doing things because it involved creating some class definition that i would then have to propagate through the headers, and wound up #defining a couple of things to make it work without that.    but it still needs to include the file and the build *should* be copying it to \orangec\include\support\win32 to be available.    Again this seems to be happening properly on my computer....

lets try to figure out what is up with the include files apparently not being copied correctly before we attempt to discover what is wrong with __fake_bind....   I tried deleting all the subdirectories in the include directory and rebuilding and it did create the subdirs again and copy files to them properly...    so I'm not sure what is up on your system...

what version of  MSVC are you using?   And are you using the default path of C:\orangec?    The build may have issues if you try to use a different path, not sure as I've never tried it...   Also, you don't have multiple versions of orangec on the path do you?   I could see all sorts of issues if you aren't running out of the same directory you build to....

**Comment by chuggafan** [comment]
I am using the default path and VS2019, and I'm building in C:\OrangeC

Though years ago I did mention doing it in a user directory, the main thing that doesn't work then is the DISTRIBUTE command, and, I believe, the installer, then again, it's been literal years since I did an out of C: build so.....

**Comment by LADSoft** [comment]
the fake bind issue was with langinfo.h not being there....    sometimes errors cascade in unexpected ways because the compiler doesn't do certain things any more once it finds an error and that just leads to other errors.    I tried to fix it last week with minimal success; will try again in a bit.

as far as that issue I had added an empty langinfo.h to address this (its contents aren't actually needed at this point although really should possibly move some stuff from xlocinfo.h)   but didn't check it in...   should work better now that I checked it in.    We'll address it and have a real langinfo.h in #434 which is in the next milestone.

**Comment by LADSoft** [comment]
now the stuff in the tests directory compiles and runs...   had to make a couple of changes due to new compiler errors and so forth...    next step is to get the compiler to compile itself :smile:

**Comment by LADSoft** [comment]
making progress on this.   Close to the end of it, although there are a couple of issues I temporarily tabled while trying to get through it that I have to revisit...

**Comment by LADSoft** [comment]
so i got it to compile itself against libcxx 10,  but it doesn't work so well; from what I can tell some non-const variable is getting included in  the const segment and a subsequent modification is causing an exception.   Will fix soon...

**Comment by LADSoft** [comment]
i did get the compiler to compile stuff last week... after compiling it with itself.   So we may be past the 'const' issue...   which was really an optimization issue, occ optimized a call statement into its function address.    

And it does seem to compile itself properly.    Right now the main impediment to closing this issue is some problem with invokables...   the compiler doesn't compile the test for invokability correctly in all instances.   I'm hoping to resolve that tonite and with any luck will be able to close this out over the weekend...

**Comment by chuggafan** [comment]
ayyy, that's great, then we can get to work on adding features so that we minimize the number of libcxx headers we override at all.

**Comment by LADSoft** [comment]
yeah looking forward to having this behind us.   Next up is the overhaul to the exception handling and typeinfo data!

**Comment by LADSoft** [comment]
more issues with out of memory when compiling problematic files....   guess I need to figure out why it is trying to use so much memory to begin with rather than trying to bandaid it again....

**Comment by LADSoft** [comment]
i found some ways to speed up compiles of the most problematic files by about 40% and use a little less memory by about 15-20% but it highlighted another bug...   simple enough to fix (i think) but I'm going to wait until tomorrow :smile:

**Comment by LADSoft** [comment]
it compiles itself again, main problem right now is that omake doesn't work reliablby when compiled with occ...   hopefully once I get past this one I can get into final testi of the build...

**Comment by GitMensch** [comment]
What do you think about moving some issues to the next milestone? iI this one is solved and possibly #523 it may even possible to get #559 to work... but in any case I _think_ it would be reasonable to have the first 2021 release "sooner". Do you agree (to move unrelated issues to the next milestone)?

**Comment by chuggafan** [comment]
AFAICT if there has been no new issues with #523 it can be closed, I'm just at this point unsure if there's been enough testing that it works reliably and we need to go into real-world compilations, though if the compiler compiles itself it should be fine to close it....

**Comment by LADSoft** [comment]
i could just have a release in the middle of a milestone lol...   but I am getting tired of looking at milestone 3 anyway.    I'll think about moving them...

**Comment by LADSoft** [comment]
as per a comment in this commit: 5df35379a4a59f618ebb0a3f6363ec44fcbf63d4 I made some changes to the source code, to use unordered maps instead of maps when reasonable.    Which was fine, the compiler still compiles stuff, just not itself.    I looked into it and it is a preexisting bug not directly related to templates...   will fix tomorrow as I want a break tonite...    there are also a couple of problems when compiling via assembly and compiling against LSCRTL.DLL that I have to address yet...   other than all that things are going pretty well :smile:

**Comment by LADSoft** [comment]
got through everything on this; now we are just pending the results of the appveyor build.

**Comment by LADSoft** [comment]
@GitMensch, I moved all issues unrelated to LIBCXX out of milestone 3.   Whenever the appveyor build goes through ok I will have a release soon after...   then finish off milestone 3 and have another release after that.

I hope development will go a little faster now, I have real problems any time I have to make major updates to the template engine and things just drag on and on.    LIBCXX tends to stress it to the max...   I wonder if they deliberately try to use every possible thing you can do in C++?   lol.    Hopefully i won't have to touch that so much and we can expedite the remaining changes...

**Comment by LADSoft** [comment]
not having much luck speeding it up further.   Tried an optimization meant to use less memory and it compiled slightly faster at home, but no difference on appveyor.   

Since we didn't have this problem last time we compiled LIBCXX8 I'm assuming it is just having problems with the new library...   debating on whether it is worth it to restore an old version of the library and see.   Maybe I will do that for completeness....   meanwhile there is some problem compiling via assembly, when compiled via assembly it doesn't assembly 'lea eax,[esp]' the same way the compiler assembled it, and that throws off all the file comparisons.   Maybe tomorrow...

**Comment by LADSoft** [comment]
just from first impressions the libcxx 8 version is going to compile faster than the libcxx 10 version...   trying to work out recently introduced bugs that keep the libcxx 8 version from fully compiling right now though...

if that is truly the case I'm going to add more of the libcxx intrinsics to see what happens, especially the ones needed by tuples as I've noticed libcxx 10 is heavy on wanting to implement things with tuples...

**Comment by chuggafan** [comment]
Any news on this? milestone 3 has been unworked on for a long time and at this point I'm scared to do work on the compiler in case that there's a ton of changes in waiting.

**Comment by LADSoft** [comment]
i haven't worked on it in a while...   When I left it I'd been trying to add two builtin templates to maybe speed up compiles of the 'tuple' portion of libcxx10...   seems reasonable both because that portion of the library seems heavily used and CLANG does have those builtin templates, probably for a reason.    But I lost some unrelated changes during an edit session and need to go back and reimplement only the necessary parts of what I was working on and then get it working.    So there aren't any outstanding changes at this point, at least on my end.

**Comment by chuggafan** [comment]
Ok sounds good, I will likely begin work on some new stuff related to this sometime next month as I go from low-to-high-to-low schedules in terms of workload.

**Comment by LADSoft** [comment]
This is coming along slowly, the rework was better than before but there are still ancilliary issues that were caused by the change.   Once we get past this we'll re-evaluate where we are with build timing and then see if it is possible to have a release.

**Comment by LADSoft** [comment]
well the builtins helped a bit on clang and gcc...  they cut out about 5 minutes.   Didn't really change the shape of microsoft builds though.   It gives me hope that a significant problem in 32-bit builds is the  wasted use of long long to be addressed in #571.   Another issue I've already started working on is the removal of calls to empty constructors.... which were happening any time a trivial structure instance was passed by value in a function call.   After looking at that for a while I think that fixing #576 is what should happen next...   because once that is fixed  I think the compiler can start omitting at least some of the exception frame setups filtered throughout various functions and I expect that would speed things up a little too.

**Comment by LADSoft** [comment]
well i took it is far as I could, trimmed off a couple of minutes but still the via assembly and withdebug builds take too long.
i may comment out those parts of the build for a while just to get it building again.   Then we can have a release with the current sources.   I profiled a little bit and likely the next step I should take is replace the inner representation of template bodies (which hold lexer tokens to be reparsed) with a statement tree...   that will make regenerating function and structure bodies more efficient.   But it is deferred until I make progress on #248...   I'm also halfway wondering if I'm attacking the right problem, it crosses my mind the appveyor may be i/o bound.   Maybe I will also do some experiments on appveyor with memory mapped files and see if that improves anything...   moving oft-used headers to the paging file might make a difference, dunno.

**Comment by LADSoft** [comment]
the build succeeded today so I'm calling this done.
